crates/ruff_linter/src/rules/pycodestyle/rules/blank_lines.rs

```diff
@@ -0,0 +1,732 @@
+                        .to_string()
+                        .repeat((BLANK_LINES_TOP_LEVEL - line.blank_lines) as usize),
+                    locator.line_start(line.first_token_range.start()),
+                )));
+
+                diagnostics.push(diagnostic);
+            }
+
+            if line.preceding_blank_lines == 0
+            // Only apply to nested functions.
+                && matches!(self.fn_status, Status::Inside(_))
+                && is_top_level_token_or_decorator(&line.first_token)
+                // Allow following a decorator (if there is an error it will be triggered on the first decorator).
+                && !matches!(self.follows, Follows::Decorator)
+                // The class's docstring can directly precede the first function.
+                && !matches!(self.follows, Follows::Docstring)
+                // Do not trigger when the def/class follows an "indenting token" (if/while/etc...).
+                && prev_indent_level.is_some_and(|prev_indent_level| prev_indent_level >= line.indent_level)
+                // Allow groups of one-liners.
+                && !(matches!(self.follows, Follows::Def) && !matches!(line.last_token, Tok::Colon))
+            {
+                // E306
+                let mut diagnostic = Diagnostic::new(
+                    BlankLinesBeforeNestedDefinition(line.blank_lines),
+                    line.first_token_range,
+                );
+
+                diagnostic.set_fix(Fix::safe_edit(Edit::insertion(
+                    stylist.line_ending().as_str().to_string(),
+                    locator.line_start(line.first_token_range.start()),
+                )));
+
+                diagnostics.push(diagnostic);
+            }
+        }
+
+        match line.first_token {
+            Tok::Class => {
+                if matches!(self.class_status, Status::Outside) {
+                    self.class_status = Status::Inside(line.indent_level + indent_size);
+                }
+                self.follows = Follows::Other;
+            }
+            Tok::At => {
+                self.follows = Follows::Decorator;
+            }
+            Tok::Def | Tok::Async => {
+                if matches!(self.fn_status, Status::Outside) {
+                    self.fn_status = Status::Inside(line.indent_level + indent_size);
+                }
+                self.follows = Follows::Def;
+            }
+            Tok::Comment(_) => {}
+            _ => {
+                self.follows = Follows::Other;
+            }
+        }
+
+        if !line.is_comment_only {
+            if !self.is_not_first_logical_line {
+                self.is_not_first_logical_line = true;
+            }
+
+            if line.is_docstring {
+                self.follows = Follows::Docstring;
+            }
+
+            self.last_non_comment_line_end = line.last_token_range.end();
+
+            if line.indent_level == 0 && !matches!(line.first_token, Tok::Comment(_)) {
```

Comment:
Testing for comments here seems unnecessary because a logical line starting with a comment is guaranteed to be a comment only line
```suggestion
            if line.indent_level == 0 {
```

Reply:
Done in 894bc21ef